### ISSUE TITLE
Maybe I'm training this wrongly??

### ISSUE THREAD

**Issue by plattsad** [AlexanderKostyak/XOJO-Machine-Learning#3]
Hi

I've tried what I think should be the most simple NN training set possible, but I'm not getting anything that looks like learning in the output.  

I wonder if you can provide any guidance?

my training data consists of 8 inputs.  The pattern 5 5 5 5 0 0 0 0 always generates an output of 1 and 0 0 0 0 5 5 5 5 always generates an output of 0.

So my input file looks like:

topology: 8 35 20 10 1
in: 0 0 0 0 5 5 5 5
out: 0
in: 5 5 5 5 0 0 0 0
out: 1

(repeated for 20,000 training samples)

But when I look at the outputs, I don't seem to get any convergence into two states and both targets seem to generate outputs that hover around the same values:

Pass19500: Inputs : 0 0 0 0 5 5 5 5 
Outputs: 0.668788 
Targets: 0 
Net recent average error: 0.497029

Pass19600: Inputs : 5 5 5 5 0 0 0 0 
Outputs: 0.953123 
Targets: 1 
Net recent average error: 0.492426

Pass19700: Inputs : 5 5 5 5 0 0 0 0 
Outputs: 0.883613 
Targets: 1 
Net recent average error: 0.507863

Pass19800: Inputs : 0 0 0 0 5 5 5 5 
Outputs: 0.972896 
Targets: 0 
Net recent average error: 0.5455

Pass19900: Inputs : 0 0 0 0 5 5 5 5 
Outputs: 0.803047 
Targets: 0 
Net recent average error: 0.576418

Pass20000: Inputs : 0 0 0 0 5 5 5 5 
Outputs: 0.699127 
Targets: 0 
Net recent average error: 0.50832

Maybe I'm misunderstanding what this aims to achieve or how I should organize the training data?

**Comment by AlexanderKostyak** [comment]
Greetings, thank you for your detailed inquiry.

The main issue you are having is because you are inputting values of 5, and the NN only supports values **0 to 1.**  The task of **normalizing** the input values, among other hyperparameter optimizations, is not included in this basic conceptualization.  Another issue you might come across is optimizing the topology of the network.  You show a topology of 8 35 20 10 1, a total of 5 layers of neurons.  This constitutes deep learning, you only need 2 or 3 layers, in fact a topology of 8 1 (8 neurons in the input and 1 output) would probably be ideal.

Some python libraries include optimization algorithms that use brute force compute to find ideal numbers for hyperparameters and coefficients for normalization (scaling) of inputs and output values, etc.  Like some forum users have stated, re-writing the wheel is not always practical.  If XOJO launched support for parallel programming in GPUs, I would strongly consider it, as their interface and language are definitely great.

Let me know if you have any continued issues.  I ran the below trainingData.txt file with quick and accurate convergence:
topology: 8 1
in: 0 0 0 0 0.5 0.5  0.5  0.5
out: 0
in: 0.5  0.5  0.5  0.5  0 0 0 0
out: 1
etc, etc, etc, 😊

**Comment by plattsad** [comment]
Thanks @AlexanderKostyak ... yes, I did try a shallower model but with no big impact. But I didn't try standardizing  - I'd kind of expected something like max(range_limit_low,min(range_limit_high,value)), but never checked for this.  So, this looks great if the issue is as simple as standardization - the above was my test set, my actual data (gene expression) is considerably more noisy and will need to be log'd or binarized to get it into range.
